### PR TITLE
[NUI] Modify CA2000 Warnings for VisualMap

### PR DESCRIPTION
**Pull request by zhouleonlei**
### Description of Change ###
Modify CA2000 Warnings for VisualMap

### API Changes ###

 - ACR:

### CODE REVIEW

**Review by elishateng — APPROVED** [review]
LGTM

**Review by AchoWang — APPROVED** [review]
Looks good to me.

**Review by rabbitfor — APPROVED** [review]
LGTM

File: src/Tizen.NUI/src/public/VisualMaps.cs [hunk 1]
```diff
@@ -660,27 +660,85 @@ protected virtual void ComposingPropertyMap()
...
+            if (visualSize != null)
+            {
+                PropertyValue temp = new PropertyValue(visualSize);
+                visualTransformMap.Add((int)VisualTransformPropertyType.Size, temp);
+                temp.Dispose();
+            }
+            if (visualOffset != null)
+            {
+                PropertyValue temp = new PropertyValue(visualOffset);
+                visualTransformMap.Add((int)VisualTransformPropertyType.Offset, temp);
```
Comment (dongsug-song) [inline src/Tizen.NUI/src/public/VisualMaps.cs:713]: could you check if the `temp` is copied when `.Add()` is called?
I am concerned about that `visualTransfromMap` is used in other place and if `temp` is disposed right after `Add()`, then the `visualTransformMap` could have null `ProperyValue` inside. 
how do you think about?

File: src/Tizen.NUI/src/public/VisualMaps.cs [hunk 2]
```diff
@@ -660,27 +660,85 @@ protected virtual void ComposingPropertyMap()
...
+                temp.Dispose();
+            }
+            if (visualOffsetPolicy != null)
+            {
+                PropertyValue temp = new PropertyValue(visualOffsetPolicy);
+                visualTransformMap.Add((int)VisualTransformPropertyType.OffsetPolicy, temp);
+                temp.Dispose();
```
Comment (dongsug-song) [inline src/Tizen.NUI/src/public/VisualMaps.cs:720]: same as above

File: src/Tizen.NUI/src/public/VisualMaps.cs [hunk 3]
```diff
@@ -660,27 +660,85 @@ protected virtual void ComposingPropertyMap()
...
+            }
+            if (visualSizePolicy != null)
+            {
+                PropertyValue temp = new PropertyValue(visualSizePolicy);
+                visualTransformMap.Add((int)VisualTransformPropertyType.SizePolicy, temp);
+                temp.Dispose();
```
Comment (dongsug-song) [inline src/Tizen.NUI/src/public/VisualMaps.cs:726]: same as above

File: src/Tizen.NUI/src/public/VisualMaps.cs [hunk 4]
```diff
@@ -660,27 +660,85 @@ protected virtual void ComposingPropertyMap()
...
+            {
+                PropertyValue temp = new PropertyValue((int)visualOrigin);
+                visualTransformMap.Add((int)VisualTransformPropertyType.Origin, temp);
+                temp.Dispose();
+            }
+            if (visualAnchorPoint != null)
+            {
+                PropertyValue temp = new PropertyValue((int)visualAnchorPoint);
+                visualTransformMap.Add((int)VisualTransformPropertyType.AnchorPoint, temp);
+                temp.Dispose();
```
Comment (dongsug-song) [inline src/Tizen.NUI/src/public/VisualMaps.cs:738]: same as above

**Review by huayongxu — COMMENTED** [review]
lgtm

**Comment by rabbitfor** [comment]
@dongsug-song ,
I've checked it.
When NUI creates PropetyValue, the corresponding dali object is created.
And then when adding it to the PropertyMap, it copies the dali object (not reference copy) in the dali code.
So I think disposing it after `Add` is safe.

File: src/Tizen.NUI/src/public/VisualMaps.cs (hunk 1 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/public/VisualMaps.cs:713]: > could you check if the `temp` is copied when `.Add()` is called?
> I am concerned about that `visualTransfromMap` is used in other place and if `temp` is disposed right after `Add()`, then the `visualTransformMap` could have null `ProperyValue` inside.
> how do you think about?
I have checked the Dali codes:
 Dali::Property::Value *arg3 = 0 ;
result = (Dali::Property::Map *) &(arg1)->Add(arg2,(Dali::Property::Value const &)*arg3);
The arg3 is the pointer of ProperyValue, it did value copied when Adding. So it is safe as Ms.Yang said.

**Comment by zhouleonlei** [comment]
> @dongsug-song ,
> I've checked it.
> When NUI creates PropetyValue, the corresponding dali object is created.
> And then when adding it to the PropertyMap, it copies the dali object (not reference copy) in the dali code.
> So I think disposing it after `Add` is safe.

Thanks for Ms.Yang careful review~

**Comment by rabbitfor** [comment]
> > @dongsug-song ,
> > I've checked it.
> > When NUI creates PropetyValue, the corresponding dali object is created.
> > And then when adding it to the PropertyMap, it copies the dali object (not reference copy) in the dali code.
> > So I think disposing it after `Add` is safe.
> 
> Thanks for Ms.Yang careful review~


Mr. Lei, could you double check?
I checked it too long ago. It is necessary to check again with the latest code.

**Comment by zhouleonlei** [comment]
> > > @dongsug-song ,
> > > I've checked it.
> > > When NUI creates PropetyValue, the corresponding dali object is created.
> > > And then when adding it to the PropertyMap, it copies the dali object (not reference copy) in the dali code.
> > > So I think disposing it after `Add` is safe.
> > 
> > 
> > Thanks for Ms.Yang careful review~
> 
> Mr. Lei, could you double check?
> I checked it too long ago. It is necessary to check again with the latest code.

I have checked Dail codes and make a Sample below:
            PropertyMap _outputVisualMap = new PropertyMap();
            PropertyValue temp = new PropertyValue("123");
            _outputVisualMap.Add(Visual.Property.Shader, temp);
            temp.Dispose();
            temp = new PropertyValue("abc");
            _outputVisualMap.Add(Visual.Property.PremultipliedAlpha, temp);
            temp.Dispose();
            for (uint i = 0; i < _outputVisualMap.Count(); i++)          
            {
                string str;
                PropertyValue v = _outputVisualMap.GetValue(i);
                v.Get(out str);
                Console.WriteLine("~~index{0} = {1}~~", i, str);
            }

It works fine. Thank you

**Pull request merged**